File: backend/services/kritiker_service.py
from sqlalchemy.orm import Session

from models import Kritiker
from models.kritiker import Kritiker
from typing import List, Optional

class KritikerService:
    def __init__(self, db: Session):
        self.db = db
    
    def get_all(self) -> list[type[Kritiker]]:
        return self.db.query(Kritiker).all()
    
    def get_by_id(self, kritiker_id: int) -> Optional[Kritiker]:
        return self.db.query(Kritiker).filter(Kritiker.kritikerid == kritiker_id).first()

    def get_by_kunden_id(self, kunden_id: int) -> Optional[Kritiker]:
        return self.db.query(Kritiker).filter(Kritiker.kundenid == kunden_id).first()
    
    def create(self, kritiker_data: dict) -> Kritiker:
        kritiker = Kritiker(**kritiker_data)
        self.db.add(kritiker)
        self.db.commit()
        self.db.refresh(kritiker)
        return kritiker
    
    def update(self, kritiker_id: int, update_data: dict) -> Optional[Kritiker]:
        kritiker = self.get_by_id(kritiker_id)
        if not kritiker:
            return None
        
        for key, value in update_data.items():
            if value is not None:  # Only update fields that are provided
                setattr(kritiker, key, value)
        
        self.db.commit()
        self.db.refresh(kritiker)
        return kritiker
    
    def delete(self, kritiker_id: int) -> bool:
        kritiker = self.get_by_id(kritiker_id)
        if not kritiker:
            return False
        self.db.delete(kritiker)
        self.db.commit()
        self.db.refresh(kritiker)
        return True

File: frontend/src/services/kritikerService.js
// services/kritikerService.js
import apiClient from '../api/apiClient';

const kritikerService = {
  /**
   * Alle Kritiker abrufen
   * GET /api/kritikers
   */
  getAll: async () => {
    return await apiClient.get('/api/kritikers');
  },

  /**
   * Einen Kritiker nach ID abrufen
   * GET /api/kritikers/{id}
   */
  getById: async (kritikerId) => {
    return await apiClient.get(`/api/kritikers/${kritikerId}`);
  },

  /**
   * Kritiker eines Kunden abrufen
   * GET /api/kritikers/by-kunden/{kundenId}
   */
  getByKundenId: async (kundenId) => {
    return await apiClient.get(`/api/kritikers/by-kunden/${kundenId}`);
  },

  /**
   * Neuen Kritiker erstellen
   * POST /api/kritikers
   */
  create: async (kritikerData) => {
    return await apiClient.post('/api/kritikers', kritikerData);
  },

  /**
   * Kritiker aktualisieren
   * PUT /api/kritikers/{id}
   */
  update: async (kritikerId, kritikerData) => {
    return await apiClient.put(`/api/kritikers/${kritikerId}`, kritikerData);
  },

  /**
   * Kritiker löschen
   * DELETE /api/kritikers/{id}
   */
  delete: async (kritikerId) => {
    return await apiClient.delete(`/api/kritikers/${kritikerId}`);
  },
};

export default kritikerService;
